constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/IntervalMap.java

```diff
@@ -0,0 +1,437 @@
+
+        index += 1;
+      }
+    }
+
+    // Cases: --<---[---]--->--
+    while (index < this.segments.size() && !IntervalAlgebra.endsAfter(this.getInterval(index), interval)) {
+      this.segments.remove(index);
+    }
+
+    // Cases: --<---[--->---]--
+    if (index < this.segments.size() && !IntervalAlgebra.startsStrictlyAfter(this.getInterval(index), interval)) {
+      // If the intervals agree on their value, we can unify the old interval with the new one.
+      // Otherwise, we'll snip the old one.
+      if (Objects.equals(this.getValue(index), value)) {
+        interval = IntervalAlgebra.unify(this.segments.remove(index).interval(), interval);
+      } else {
+        final var suffix = IntervalAlgebra.intersect(this.getInterval(index), IntervalAlgebra.upperBoundsOf(interval));
+
+        this.segments.set(index, Segment.of(suffix, this.getValue(index)));
+      }
+    }
+
+    // now, everything left of `index` is strictly left of `interval`,
+    // and everything right of `index` is strictly right of `interval`,
+    // so adding this interval to the list is trivial.
+    this.segments.add(index, Segment.of(interval, value));
+
+    return index;
+  }
+
+  public IntervalMap<V> unset(final Interval interval) {
+    return unset(new IntervalMap<>(interval, new Object()));
+  }
+
+  public IntervalMap<V> unset(final List<Interval> intervals) {
+    return unset(new IntervalMap<>(intervals, new Object()));
+  }
+
+  public IntervalMap<V> unset(final IntervalMap<?> other) {
+    return map2(
+        this, other,
+        (l, r) -> r.isPresent() ? Optional.empty() : l
+    );
+  }
+
+  public IntervalMap<V> select(final Interval bounds) {
+    return unset(List.of(
+        interval(Duration.MIN_VALUE, Inclusive, bounds.start, bounds.startInclusivity.opposite()),
+        interval(bounds.end, bounds.endInclusivity.opposite(), Duration.MAX_VALUE, Inclusive)
+    ));
+  }
+
+  public IntervalMap<V> select(final List<Interval> intervals) {
+    return map2(
+        this, new IntervalMap<>(intervals, new Object()),
+        (l, r) -> r.isPresent() ? l : Optional.empty()
+    );
+  }
+
+  /**
+   * Authored by Jonathan
+   * Maps intervals and the gaps between them in IntervalMap intervals to new values following some function transform
+   *  which converts the old values (or nulls, in the case of gaps) to new values.
+   *
+   * @param transform the function to apply to each value of intervals (or the gaps between!), mapping its original value type V to another value type R
+   * @return a new IntervalMap with newly mapped values
+   * @param <R> The new value type that the returned IntervalMap's intervals should correspond to
+   */
+  public <R> IntervalMap<R> map(final Function<Optional<V>, Optional<R>> transform) {
+    return this.contextMap((v, i) -> transform.apply(v));
+  }
+
+  public <R> IntervalMap<R> contextMap(final BiFunction<Optional<V>, Interval, Optional<R>> transform) {
```

Comment:
Technically that would behave differently if the transform had side-effects, but that's not a good use case. I've removed `insertAndCoalesce`.

Personally I'd like the symmetry that `mapN` performs a transform with signature `Optional<?> ^ N -> Optional<R>`, that's my only objection.

Reply:
(Yeah, I see what you're saying.)

All non-test uses of `map2` that I could find also send `empty() * empty()` to `empty()`. I consider this a well-foundedness criterion, anyway -- _mumble mumble_ category of [based sets](https://en.wikipedia.org/wiki/Pointed_set) and base-preserving functions _mumble mumble_.

This is in fact why I defined `BinaryTransformation` in [my earlier prototypes](https://gist.github.com/Twisol/88a8b2bf9e71681e74c1b4160ee9a03e) -- it lets us split the cases in a `map2` combiner into three, making it impossible to get the fourth wrong, where both arguments are empty.

Reply:
lol. Based set = (paraphrase) "A set with an element that is special. Its not special because it has any unique properties; its just special because I said so" :)

Humanity has officially done Too Much Math 

Ok I'll try it out and see how it goes

Reply:
lol yes, i was very surprised when i realized it had any relevance at all to us, but sometimes math maths its way into things.

Reply:
Oh yes, the precise relevance is here:

> The category of pointed sets and based maps is equivalent to the category of sets and [partial functions](https://en.wikipedia.org/wiki/Partial_function). The base point serves as a "default value" for those arguments for which the partial function is not defined.

Reply:
map is converted to what you suggested, map2 is not yet.